File: src/libs/base/arch/gem5/Backtrace.cc
/*
 * Copyright (C) 2015, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/util/Math.h>
#include <base/Backtrace.h>
#include <base/Config.h>

namespace m3 {

const size_t Backtrace::CALL_INSTR_SIZE = 5;

size_t Backtrace::collect(uintptr_t *addr, size_t max) {
    uintptr_t bp;
    asm volatile ("mov %%rbp,%0" : "=a" (bp));

    uintptr_t stack = Math::round_dn<uintptr_t>(bp, STACK_SIZE);
    uintptr_t end = Math::round_up<uintptr_t>(bp, STACK_SIZE);
    uintptr_t start = end - STACK_SIZE;
    size_t i = 0;
    for(; i < max; ++i) {
        // prevent page-fault
        if(bp < start || bp >= end)
            break;

        bp = stack + (bp & (STACK_SIZE - 1));
        addr[i] = *(reinterpret_cast<uintptr_t*>(bp) + 1) - CALL_INSTR_SIZE;
        bp = *reinterpret_cast<uintptr_t*>(bp);
    }
    return i;
}

}

File: src/libs/rust/src/backtrace.rs
use cfg;
use util;

fn get_bp() -> usize {
    let val: usize;
    unsafe {
        asm!(
            "mov %rbp, $0"
            : "=r"(val)
        );
    }
    val
}

pub fn collect(addr: &mut [usize]) -> usize {
    let mut bp = get_bp();

    let base = util::round_dn(bp, cfg::STACK_SIZE);
    let end = util::round_up(bp, cfg::STACK_SIZE);
    let start = end - cfg::STACK_SIZE;

    for i in 0..addr.len() {
        if bp < start || bp >= end {
            return i;
        }

        bp = base + (bp & cfg::STACK_SIZE - 1);
        let bp_ptr = bp as *const usize;
        unsafe {
            addr[i] = *bp_ptr.offset(1) - 5;
            bp = *bp_ptr;
        }
    }
    return addr.len();
}
